lib/sources.nix

```diff
@@ -101,4 +101,201 @@ rec {
   pathHasContext = builtins.hasContext or (lib.hasPrefix builtins.storeDir);
 
   canCleanSource = src: src ? _isLibCleanSourceWith || !(pathHasContext (toString src));
+
+  # Splits a filesystem path into its components.
+  splitPath = path: lib.splitString "/" (toString path);
+
+  # Turns a list of path components into a tree, e.g.
+  #
+  #     ["a" "b" "c1"]
+  #     ["a" "b" "c2"]
+  #     ["a" "b" "c3"]
+  #     ["a" "x"     ]
+  #
+  # becomes:
+  #
+  #     { a = { b = { c1 = null; c2 = null; c3 = null }; x = null; }; }
+  pathComponentsToTree = paths: with lib;
+    foldl (tree: path: recursiveUpdate tree (setAttrByPath path null)) {} paths;
+
+  # Returns true if and only if any prefix of the given `path` leads to a leaf
+  # (`null`) in the given `tree` (nested attrset).
+  # That is: "If we go down the tree by the given path, do we hit a leaf?"
+  #
+  # Example: For the tree
+  #
+  #     a
+  #       b-leaf
+  #       c-leaf
+  #       d
+  #         e-leaf
+  #     f-leaf
+  #
+  # represented as attrset
+  #
+  #     { a = { b = null; c = null; d = { e = null; }; }; f = null; }
+  #
+  # we have
+  #
+  #     isPrefixOfLeafPath [a] == true
+  #     isPrefixOfLeafPath [x] == false
+  #     isPrefixOfLeafPath [a b] == true
+  #     isPrefixOfLeafPath [a b c] == true
+  #     isPrefixOfLeafPath [a b c x] == true
+  #     isPrefixOfLeafPath [a d] == false
```

Comment:
I don't get this function. Shouldn't `isPrefixOfLeafPath [a d] == true`? Because `isPrefixOfLeafPath [a] == true` and `[a]` is a prefix of `[a d]`.

Reply:
Sorry, `isPrefixOfLeafPath [a] == false`; that was a typo.